h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+    // next time they create an annotation.
+    // But _don't_ cache it when they change the privacy level of a reply.
+    if (!model.references) {  // If the annotation is not a reply.
+      permissions.setDefault(privacy);
+    }
+    if (privacy === 'private') {
+      vm.annotation.permissions = permissions.private();
+    } else if (privacy === 'shared') {
+      vm.annotation.permissions = permissions.shared(model.group);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotaitonController#hasContent
+    * @returns {boolean} True if the currently edited annotation has
+    *          content (ie. is not just a highlight)
+    */
+  vm.hasContent = function() {
+    var textLength = (vm.annotation.text || '').length;
+    var tagsLength = (vm.annotation.tags || []).length;
+    return (textLength > 0 || tagsLength > 0);
+  };
+
+  /**
+    * @returns {boolean} True if this annotation has quotes
+    */
+  vm.hasQuotes = function() {
+    return vm.annotation.target.some(function(target) {
+      return target.selector && target.selector.some(function(selector) {
+        return selector.type === 'TextQuoteSelector';
+      });
+    });
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#authorize
+    * @param {string} action The action to authorize.
+    * @returns {boolean} True if the action is authorized for the current user.
+    * @description Checks whether the current user can perform an action on
+    * the annotation.
+    */
+  vm.authorize = function(action) {
+    if (model === null) {
+      return false;
+    }
+    // TODO: this should use auth instead of permissions but we might need
+    // an auth cache or the JWT -> userid decoding might start to be a
+    // performance bottleneck and we would need to get the id token into the
+    // session, which we should probably do anyway (and move to opaque bearer
+    // tokens for the access token).
+    return permissions.permits(action, model, session.state.userid);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#delete
+    * @description Deletes the annotation.
+    */
+  vm['delete'] = function() {
+    return $timeout(function() {  // Don't use confirm inside the digest cycle.
+      var onRejected;
+      var msg = 'Are you sure you want to delete this annotation?';
+      if ($window.confirm(msg)) {
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Deleting annotation failed');
+        };
+        $scope.$apply(function() {
+          annotationMapper.deleteAnnotation(model).then(
+            null, onRejected);
+        });
+      }
+    }, true);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#edit
+    * @description Switches the view to an editor.
+    */
+  vm.edit = function() {
+    if (!drafts.get(model)) {
+      updateDraft(model);
+    }
+    vm.action = model.id !== null ? 'edit' : 'create';
```

Comment:
How do you feel about these ternary expressions anyway? I find them really hard to read, would rather have a bigger `if` / `else`